ZuluContent/Engines/Harvest/Core/HarvestDefinition.cs

```diff
@@ -1,179 +1,323 @@
-		public bool RandomizeVeins { get { return m_RandomizeVeins; } set { m_RandomizeVeins = value; } }
-
-		private Dictionary<Map, Dictionary<Point2D, HarvestBank>> m_BanksByMap;
-
-		public Dictionary<Map, Dictionary<Point2D, HarvestBank>> Banks{ get{ return m_BanksByMap; } set{ m_BanksByMap = value; } }
-
-		public void SendMessageTo( Mobile from, object message )
-		{
-			if ( message is int )
-				from.SendLocalizedMessage( (int)message );
-			else if ( message is string )
-				from.SendMessage( (string)message );
-		}
-
-		public HarvestBank GetBank( Map map, int x, int y )
-		{
-			if ( map == null || map == Map.Internal )
-				return null;
-
-			x /= m_BankWidth;
-			y /= m_BankHeight;
-
-			Dictionary<Point2D, HarvestBank> banks = null;
-			m_BanksByMap.TryGetValue( map, out banks );
-
-			if ( banks == null )
-				m_BanksByMap[map] = banks = new Dictionary<Point2D, HarvestBank>();
-
-			Point2D key = new Point2D( x, y );
-			HarvestBank bank = null;
-			banks.TryGetValue( key, out bank );
-
-			if ( bank == null )
-				banks[key] = bank = new HarvestBank( this, GetVeinAt( map, x, y ) );
-
-			return bank;
-		}
-
-		public HarvestVein GetVeinAt( Map map, int x, int y )
-		{
-			if ( m_Veins.Length == 1 )
-				return m_Veins[0];
-
-			double randomValue;
-
-			if ( m_RandomizeVeins )
-			{
-				randomValue = Utility.RandomDouble();
-			}
-			else
-			{
-				System.Random random = new System.Random( x * 17 + y * 11 + map.MapID * 3 );
-				randomValue = random.NextDouble();
-			}
-
-			return GetVeinFrom( randomValue );
-		}
-
-		public HarvestVein GetVeinFrom( double randomValue )
-		{
-			if ( m_Veins.Length == 1 )
-				return m_Veins[0];
-
-			randomValue *= 100;
-
-			for ( int i = 0; i < m_Veins.Length; ++i )
-			{
-				if ( randomValue <= m_Veins[i].VeinChance )
-					return m_Veins[i];
-
-				randomValue -= m_Veins[i].VeinChance;
-			}
-
-			return null;
-		}
-
-		public BonusHarvestResource GetBonusResource()
-		{
-			if ( m_BonusResources == null )
-				return null;
-
-			double randomValue = Utility.RandomDouble() * 100;
-
-			for ( int i = 0; i < m_BonusResources.Length; ++i )
-			{
-				if ( randomValue <= m_BonusResources[i].Chance )
-					return m_BonusResources[i];
-
-				randomValue -= m_BonusResources[i].Chance;
-			}
-
-			return null;
-		}
-
-		public HarvestDefinition()
-		{
-			m_BanksByMap = new Dictionary<Map, Dictionary<Point2D, HarvestBank>>();
-		}
+    public delegate int ConsumedPerHarvestCallback(double skillValue);
+
+    public delegate void BonusEffectCallback(Mobile harvester, Item tool);
+
+    public class HarvestDefinition
+    {
+        private int m_BankWidth, m_BankHeight;
+        private int m_MinTotal, m_MaxTotal;
+        private int[] m_Tiles;
+        private bool m_RangedTiles;
+        private TimeSpan m_MinRespawn, m_MaxRespawn;
+        private ConsumedPerHarvestCallback m_ConsumedPerHarvestCallback;
+        private BonusEffectCallback m_BonusEffectCallback;
```

Comment:
So you don't need to declare delegates if you don't want to, a shorthand way of doing this is using the `Func<>` and `Action<>` in the System namespace.

```csharp
        private ConsumedPerHarvestCallback m_ConsumedPerHarvestCallback;
        private BonusEffectCallback m_BonusEffectCallback;
```
Becomes
```csharp
        private Func<double, int> m_ConsumedPerHarvestCallback;
        private Action<Mobile, Item> m_BonusEffectCallback;
```

Then you can save defining the delegates. The one draw back is that you lose the implied names of the parameters, so anyone calling it has less context.

Reply:
Cool thanks for pointing it out!